### PR TITLE
PERF-#4732: Avoid overwriting already-evaluated _width_cache and _length_cache

### LINKED ISSUE #4732: DOC: which classes are immutable?
Doing some performance troubleshooting I'm finding a call to `PandasOnRayDataframePartition.drain_call_queue` is overriding a already-evaluated `self._width_cache` with a ObjectIDType object, which later takes an absurd amount of time to access via ray.get.

Is it the case that a Partition's shape should not change following `__init__`?  if so, it would be nice to have that clearly documented.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Signed-off-by: Brock <jbrockmendel@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4732
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by mvashishtha — COMMENTED** [review]
@jbrockmendel This change seems fine to me, but I'm still curious to see an example where this is useful for performance.

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py [hunk 1]
```diff
@@ -170,13 +170,20 @@ def drain_call_queue(self):
...
                 self._ip_cache,
             ) = _apply_func.remote(data, func, *args, **kwargs)
         logger.debug(f"EXIT::Partition.drain_call_queue::{self._identity}")
         self.call_queue = []
 
+        # GH#4732 if we already have evaluated width/length cached as ints,
+        #  don't overwrite that cache with non-evaluated values.
+        if not isinstance(self._length_cache, int):
+            self._length_cache = new_length
+        if not isinstance(self._width_cache, int):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:184]: that would miss the case where `self._length_cache is None`

**Comment by jbrockmendel** [comment]
> but I'm still curious to see an example where this is useful for performance

I don't have a case where it makes a big difference, but that is largely because _nothing_ seems to make a big difference as long as 90% of my profiled runtime is in `ray.get`.

File: modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvashishtha) [reply modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py:184]: True. This is okay with me, then.

**Review by mvashishtha — APPROVED** [review]
LGTM

**Comment by YarShev** [comment]
@jbrockmendel, the change is probably reasonable but I am not sure it makes a big difference. I have been trying to cover most of the cases for block-wise partitions when we can preserve lengths and widths in this branch - https://github.com/YarShev/modin/tree/dev/yigoshev-caches. Would you be able to profile the branch?

**Comment by jbrockmendel** [comment]
> Would you be able to profile the branch?

Based on a cut-down version of the script we've been profiling (truncated the parquet file at 100k records) the branch you linked is quite a bit faster than master.  Measured each branch 3 times and yours took 36.1s, 25.8s, 36.4s, master took 66.0s, 58.7s,  56.8s.

That said, the branch you linked is several commits behind master, which I'm pretty sure saw a significant slowdown over the weekend.  Can you rebase and I'll re-run?

**Comment by YarShev** [comment]
@jbrockmendel, here it is https://github.com/YarShev/modin/commits/dev/yigoshev-caches-synced-master. Please note that this is a new branch.

**Comment by jbrockmendel** [comment]
> here it is https://github.com/YarShev/modin/commits/dev/yigoshev-caches-synced-master. Please note that this is a new branch.

Timings on this branch 41.7s, 40.1s, 30.2s

**Comment by YarShev** [comment]
@jbrockmendel, great, thanks! We'll probably need to cut off unnecessary changes from the branch and submit a PR. Btw, are you seeing any speed up from these changes in the script you are profiling?

**Comment by jbrockmendel** [comment]
> Btw, are you seeing any speed up from these changes in the script you are profiling?

I'm not seeing any speed up from anything.  All the time is spent inside ray.get no matter what I do.

**Comment by jbrockmendel** [comment]
Anything left to do here?

**Comment by YarShev** [comment]
Let's wait for CI to be finished. I reran the job that failed but it is not related to the changes in this PR I believe.

**Review by YarShev — APPROVED** [review]

**Pull request merged**